build-logic/build-update-utils/src/main/kotlin/gradlebuild/buildutils/tasks/AbstractCheckOrUpdateContributorsInReleaseNote.kt

```diff
@@ -0,0 +1,162 @@
+
+    @Internal
+    protected fun getContributorsInReleaseNote(): Set<GitHubUser> {
+        val releaseNoteText = releaseNote.asFile.get().readText()
+        if (!contributorSectionRegex.containsMatchIn(releaseNoteText)) {
+            throw IllegalStateException("We can't find contributors in the release note $releaseNote. Is anything changed?")
+        }
+        return contributorSectionRegex.find(releaseNoteText)!!.groupValues[2].lines()
+            .map { it.trim() }
+            .filter { it.isNotEmpty() }
+            .onEach { if (!contributorLineRegex.containsMatchIn(it)) throw IllegalStateException("Invalid contributor line: $it") }
+            .map { GitHubUser(contributorLineRegex.find(it)!!.groupValues[2], contributorLineRegex.find(it)!!.groupValues[1]) }
+            .toSet()
+    }
+
+    @Internal
+    protected fun getContributorsFromPullRequests(): Set<GitHubUser> {
+        if (!milestone.isPresent) {
+            throw IllegalStateException("Milestone not set: please rerun the task with `--milestone <milestone>`")
+        }
+        val prs: MutableList<GitHubPullRequest> = mutableListOf()
+        var page = 0
+        while (++page <= 10) { // at most 1000 PRs
+            val prPage = getMergedContributorPullRequests(page)
+            prs.addAll(prPage)
+            if (prPage.size < pageSize) {
+                break
+            }
+        }
+        return prs
+            .filter { it.milestone?.title?.startsWith(milestone.get()) == true }
+            .map { it.user.login }
+            .toSet()
+            .map { getUserInfo(it) }
+            .toSet()
+    }
+
+    private fun <T> invokeGitHubApi(uri: String, klass: Class<T>): T {
+        val request = HttpRequest.newBuilder()
+            .uri(URI(uri))
+            .apply {
+                if (githubToken.isPresent) {
+                    header("Authorization", "token ${githubToken.get()}")
+                }
+            }
+            .build()
+        val response = HttpClient.newHttpClient().send(request, HttpResponse.BodyHandlers.ofString())
+        if (response.statusCode() > 399) {
+            throw RuntimeException("Failed to get pull requests: $uri ${response.statusCode()} ${response.body()}")
+        }
+        return Gson().fromJson(response.body(), klass)
+    }
+
+    private fun getUserInfo(login: String): GitHubUser {
+        val uri = "https://api.github.com/users/$login"
+        return invokeGitHubApi(uri, GitHubUser::class.java)
+    }
+
+    private fun getMergedContributorPullRequests(pageNumber: Int): List<GitHubPullRequest> {
+        val uri = "https://api.github.com/search/issues?q=is:pr+is:merged+repo:gradle/gradle+label:%22from:contributor%22&sort=updated&order=desc&per_page=$pageSize&page=$pageNumber"
+        return invokeGitHubApi(uri, GitHubPullRequestSearchResult::class.java).items
+    }
+}
+
+open class CheckContributorsInReleaseNote : AbstractCheckOrUpdateContributorsInReleaseNote() {
```

Comment:
Can you put those two subclasses in separate files?

Reply:
Done.